services/comments.js

```diff
@@ -0,0 +1,277 @@
+        };
+        const entities = query._q ? 
+            await strapi.query( model.modelName, pluginName).search(params, ['authorUser', 'related', 'reports']) :
+            await strapi.query( model.modelName, pluginName).find(params, ['authorUser', 'related', 'reports']);
+        const items = entities.map(_ => filterOurResolvedReports(sanitizeEntity(_, { model })));
+        const total = paginationEnabled ?
+            query._q ?
+                await strapi.query( model.modelName, pluginName).countSearch(params) : 
+                await strapi.query( model.modelName, pluginName).count(params) :
+            items.length;
+        return {
+            items,
+            total,
+            page: paginationEnabled ? query._start/query._limit : undefined,
+        };
+    },
+
+    // Find comments and create relations tree
+    findAllInHierarchy: async (relation, startingFromId = null, dropBlockedThreads = false) => {
+        const [relationType, relationId] = (relation || '').split(':');
+        const { pluginName, model } = extractMeta(strapi.plugins);
+        const entities = await strapi.query( model.modelName, pluginName).find({}, ['authorUser', 'related', 'reports'])
+            .then(results => relation ? results.filter(result => {
+                    const { related } = result || {};
+                    if (related && !_.isEmpty(related)) {
+                        return _.find(related, item =>
+                                (`${item.id}` === relationId) &&
+                                (item.__contentType.toLowerCase() === relationType.toLowerCase())
+                            );
+                    }
+                    return false;
+                }) : results
+            );
+        return buildNestedStructure(entities.map(_ => filterOurResolvedReports(sanitizeEntity(_, { model }))), startingFromId, 'threadOf', dropBlockedThreads);
+    },
+
+    // Find single comment
+    findOne: async (id, relation) => {
+        const [relationType, relationId] = (relation || '').split(':');
+        const { pluginName, model } = extractMeta(strapi.plugins);
+        const entity = await strapi.query( model.modelName, pluginName).findOne({ id }, ['related', 'reports'])
+            .then(result => {
+                if (relation) {
+                    const { related } = result || {};
+                    if (related && !_.isEmpty(related)) {
+                        if (_.find(related, resultItem => (resultItem.id.toString() === relationId) && (resultItem.__contentType.toLowerCase() === relationType.toLowerCase()))) {
+                            return result;
+                        }
+                        return null;
+                    }
+                }
+                return result;
+            });
+        return filterOurResolvedReports(sanitizeEntity(entity, { model }));
+    },
+
+    // Create a comment
+    create: async (data, relation) => {
```

Comment:
why do we sometimes `throw` and sometimes `return` PluginError? it is thrown out of `checkBadWords` and returned in other cases